README.md
# rich_text_view

A simple yet powerful rich text view that supports mention, hashtag, email, url and view more.

## Example

<p>
    <img src="https://raw.githubusercontent.com/nelstein/rich_text_view/main/screenshots/homepage.png" width="200px" height="auto" hspace="20"/>
</p>

###  RichTextView 
```dart
 RichTextView(
           text:
                      '''Who else thinks it's thinks it's just cool😎 to mention 
                      @jane when #JaneMustLive is trending without even trying 
                      to send a *bold* email to janedoe@gmail.com and verify the
                       facts talkmore of visiting www.janedoe.com''',
                  maxLines: 3,
                  truncate: true,
                  viewLessText: 'less',
                  linkStyle: TextStyle(color: Colors.blue),
                  supportedTypes: [
                    EmailParser(
                        onTap: (email) => print('${email.value} clicked')),
                    PhoneParser(
                        onTap: (phone) => print('click phone ${phone.value}')),
                    MentionParser(
                        onTap: (mention) => print('${mention.value} clicked')),
                    UrlParser(onTap: (url) => print('visting ${url.value}?')),
                    BoldParser(),
                    HashTagParser(
                        onTap: (hashtag) =>
                            print('is ${hashtag.value} trending?'))
                  ]
                )
```
### RichTextEditor 

RichTextEditor is an input field that supports suggestions when  mentioning or using hashtags

```dart

        RichTextEditor(
                      decoration: InputDecoration(
                        border: OutlineInputBorder(
                          borderRadius: BorderRadius.circular(16)),
                      ),
                      suggestionController: SuggestionController(
                      mentionSymbol: '/',
                      position: SuggestionPosition.bottom,
                      mentionSuggestions: [
                        Mention(
                            imageURL: 'imageURL',
                            subtitle: 'nelly',
                            title: 'Nelly Gane'),
                        Mention(
                            imageURL: 'imageURL',
                            subtitle: 'gaus',
                            title: 'Gaus Shell')
                      ],
                      onSearchMention: (term) async {
                        return List.generate(
                            20,
                            (index) => Mention(
                                id: index.toString(),
                                imageURL: 'imageURL',
                                subtitle: term.toLowerCase(),
                                title: '$term $index'));
                      },
                      onMentionSelected: (suggestion) {
                        print(suggestion.id);
                      },
                      onSearchTags: (term) async {
                        return [
                          HashTag(
                              hashtag: '#Dart',
                              subtitle: '30 posts',
                              trending: true),
                          HashTag(
                            hashtag: '#Flutter',
                            subtitle: '56 posts',
                          )
                        ];
                      },
                    ))
```

### Todo

* Add styles to mentions/hashtags in RichTextEditor


CHANGELOG.md
##[1.3.0]

- Updated `RTUtils.emailPattern` and `RTUtils.urlPattern` to use RFC 5322 Standard Regex
- disposed `suggestionController` correctly
- added `ItalicsParser` and `viewMoreLessStyle`
- Fixed (#31)[https://github.com/nelsonweze/rich_text_view/issues/31] and (#32)[https://github.com/nelsonweze/rich_text_view/issues/32]
- Added `borderColor` property to `SuggestionController`

##[1.2.1-dev-1]

- Fixed (#31)[https://github.com/nelsonweze/rich_text_view/issues/31] and (#32)[https://github.com/nelsonweze/rich_text_view/issues/32]
- Added `borderColor` property to `SuggestionController`

## [1.2.0]

- BREAKING: Removed `ParsedType`, renamed `MatchText` to `ParserType`
- `RichTextView.supportedTypes` now takes list of `ParserType` eg `MentionParser`, `HashTagParser`, `UrlParser` and so on. You can extend `ParserType` for a custom parser
- All onTap callbacks in `RichTextView` have been moved to thier individual parser class
- `ParserType.onTap` returns `Matched` which also contains the start and end indexes
- Check example codes for more

## [1.1.0]

- stable release
- minor fixes

## [1.1.0-dev-1]

- BREAKING: Added SuggestionController which now takes more customizations
- Removed dependency on Bloc package,
- Added id param to `Mention`

## [1.0.0]

- Stable release

## [1.0.0-dev-2]

- BREAKING: Separated RichTextEditor from RichTextView to avoid confusion with parameters.
- Added more text field parameters
- Added support to toggle between view more and view less

## [1.0.0-dev-1]

- BREAKING: `Suggestion` renamed to `Mention`
  `onSearchPeople` renamed to `onSearchMention`
  `onSuggestionSelected` renamed to `onMentionSelected`
  `showMoreText` renamed to `truncate`
- Added callbacks for when a mention or hashtag is selected during search
- Added option to use custom widgets for suggestion items

## [0.0.3]

- Fix bold style issue
- Fix issue when selectable is set to false

## [0.0.2]

- To use as a TextField, there's now RichTextView.editor
- Added screenshot to README.md
- fixed some bugs and improved example codes

## [0.0.1]

- Initial release to pub.dev


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:rich_text_view/rich_text_view.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
        title: 'Flutter Demo',
        theme: ThemeData(
          primarySwatch: Colors.blue,
        ),
        home: HomePage());
  }
}

class HomePage extends StatelessWidget {
  const HomePage({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('Flutter RichTextView'),
      ),
      body: SingleChildScrollView(
        child: Column(
          children: [
            SizedBox(
              height: 32,
            ),
            Text('RichTextView'),
            SizedBox(
              height: 16,
            ),
            Container(
              alignment: Alignment.center,
              padding: EdgeInsets.symmetric(horizontal: 16),
              child: RichTextView(
                text:
                    '''Who else thinks it's thinks it's just cool😊 to mention 
                      @jane when #JaneMustLive is trending without even trying 
                      to send a *bold* email to janedoe@gmail.com and verify the
                       facts talkmore of ivisitingi www.janedoe.com''',
                maxLines: 3,
                truncate: true,
                viewLessText: 'less',
                linkStyle: TextStyle(color: Colors.blue),
                textAlign: TextAlign.center,
                supportedTypes: [
                  EmailParser(
                      onTap: (email) => print('${email.value} clicked')),
                  PhoneParser(
                      onTap: (phone) => print('click phone ${phone.value}')),
                  MentionParser(
                      onTap: (mention) => print('${mention.value} clicked')),
                  UrlParser(onTap: (url) => print('visting ${url.value}?')),
                  BoldParser(),
                  HashTagParser(
                      onTap: (hashtag) =>
                          print('is ${hashtag.value} trending?'))
                ],
              ),
            ),
            SizedBox(
              height: 48,
            ),
            Text('RichTextEditor'),
            SizedBox(
              height: 16,
            ),
            SizedBox(
              width: 300,
              child: RichTextEditor(
                  decoration: InputDecoration(
                    border: OutlineInputBorder(
                        borderRadius: BorderRadius.circular(16)),
                  ),
                  suggestionController: SuggestionController(
                    mentionSymbol: '/',
                    position: SuggestionPosition.bottom,
                    mentionSuggestions: [
                      Mention(
                          imageURL: 'imageURL',
                          subtitle: 'nelly',
                          title: 'Nelly Gane'),
                      Mention(
                          imageURL: 'imageURL',
                          subtitle: 'gaus',
                          title: 'Gaus Shell')
                    ],
                    onSearchMention: (term) async {
                      return List.generate(
                          20,
                          (index) => Mention(
                              id: index.toString(),
                              imageURL: 'https://via.placeholder.com/150',
                              subtitle: term.toLowerCase(),
                              title: '$term $index'));
                    },
                    onMentionSelected: (suggestion) {
                      print(suggestion.id);
                    },
                    onSearchTags: (term) async {
                      return [
                        HashTag(
                            hashtag: '#Dart',
                            subtitle: '30 posts',
                            trending: true),
                        HashTag(
                          hashtag: '#Flutter',
                          subtitle: '56 posts',
                        )
                      ];
                    },
                  )),
            ),
          ],
        ),
      ),
    );
  }
}
